src/lxc/cgroups/cgfsng.c

```diff
@@ -947,6 +953,358 @@ static bool check_cgroup_dir_config(struct lxc_conf *conf)
+		return log_error_errno(false, errno, "Failed opening cgroup dir \"%s\"", parent_cgroup);
+
+	// Create new child cgroup 'init', and move the task there
+	new_scope = must_make_path(parent_cgroup, "init", NULL);
+	ret = mkdir_p(new_scope, 0755);
+	if (ret < 0 && (errno != EEXIST))
+		return false;
+	buf = read_file_at(pdir, "cgroup.procs", PROTECT_OPEN, 0);
+	if (!buf)
+		return false;
+	new_scope_pids = must_make_path(new_scope, "cgroup.procs", NULL);
+	ret = lxc_write_to_file(new_scope_pids, buf, strlen(buf), false, 0666);
+	if (ret != 0)
+		return log_error_errno(false, errno, "Failed to escape to cgroup \"%s\"", new_scope);
+
+	// enable controllers in parent_cgroup
+	return enable_controllers_delegation(pdir, parent_cgroup);
+}
+
+static int unpriv_systemd_create_scope(struct cgroup_ops *ops, struct lxc_conf *conf)
+{
+	__do_free char *full_scope_name = NULL;
+	__do_free char *fs_cg_path = NULL;
+	sd_event *event = NULL;
+	__attribute__((__cleanup__(sd_bus_unrefp))) sd_bus *bus = NULL; // free the bus before the names it references, just to be sure
+	struct sd_callback_data sd_data;
+	int idx = 0;
+	size_t len;
+	int r;
+
+	if (geteuid() == 0)
+		return log_info(SYSTEMD_SCOPE_UNSUPP, "Running privileged, not using a systemd unit");
+	// Pure_unified_layout() can't be used as that info is not yet setup.  At
+	// the same time, we don't want to calculate current cgroups until after
+	// we optionally enter a new systemd user scope.  So let's just do a quick
+	// check for pure unified cgroup system: single line /proc/self/cgroup with
+	// only index '0:'
+	if (!pure_unified_system())
+		return log_info(SYSTEMD_SCOPE_UNSUPP, "Not in unified layout, not using a systemd unit");
+
+	r = sd_bus_open_user(&bus);
+	if (r < 0)
+		return log_error(SYSTEMD_SCOPE_FAILED, "Failed to connect to user bus: %s", strerror(-r));
+
+	r = sd_bus_call_method_asyncv(bus, NULL, DESTINATION, PATH, INTERFACE, "Subscribe", NULL, NULL, NULL, NULL);
+	if (r < 0)
+		return log_error(SYSTEMD_SCOPE_FAILED, "Failed to subscribe to signals: %s", strerror(-r));
+
+	sd_data.job_complete = false;
+	sd_data.scope_name = NULL;
+	r = sd_bus_match_signal(bus,
+		NULL, // no slot
+		DESTINATION, PATH, INTERFACE, "JobRemoved",
+		systemd_jobremoved_callback, &sd_data);
+	if (r < 0)
+		return log_error(SYSTEMD_SCOPE_FAILED, "Failed to register systemd event loop signal handler: %s", strerror(-r));
+
+	// NEXT: create and attach event
+	r = sd_event_new(&event);
+	if (r < 0)
+		return log_error(SYSTEMD_SCOPE_FAILED, "Failed allocating new event: %s\n", strerror(-r));
+	r = sd_bus_attach_event(bus, event, SD_EVENT_PRIORITY_NORMAL);
+	if (r < 0) {
+		// bus won't clean up event since the attach failed
+		sd_event_unrefp(&event);
+		return log_error(SYSTEMD_SCOPE_FAILED, "Failed attaching event: %s\n", strerror(-r));
+	}
+
+	len = strlen("lxc-") + strlen(conf->name) + 4 + 6;
+	full_scope_name = malloc(len + 7);
+	if (!full_scope_name)
+		return log_error(-ENOMEM, "Out of memory");
```

Comment:
Fwiw, you can just use `syserror("Out of memory")` in such cases. It'll return a negative errno.

Reply:
Neat.